svcdiscovery/plugin/svcdiscovery/plugin.go

```diff
@@ -0,0 +1,116 @@
+/*
+Copyright 2020 SUSE
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+	http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package svcdiscovery
+
+import (
+	"context"
+	"net"
+
+	"github.com/coredns/coredns/plugin"
+	clog "github.com/coredns/coredns/plugin/pkg/log"
+	"github.com/coredns/coredns/request"
+	"github.com/miekg/dns"
+)
+
+// ServiceDiscovery is a CoreDNS plugin for handling Cloud Foundry App Service
+// Discovery.
+type ServiceDiscovery struct {
+	Next plugin.Handler
+
+	log       clog.P
+	sdcClient *SDCClient
+	ttl       uint32
+}
+
+// Name returns the plugin name.
+func (sd *ServiceDiscovery) Name() string { return pluginName }
+
+// ServeDNS handles domain name requests for Cloud Foundry App Service Discovery
+// by resolving internal domain names.
+func (sd *ServiceDiscovery) ServeDNS(
+	ctx context.Context,
+	res dns.ResponseWriter,
+	req *dns.Msg,
+) (int, error) {
+	state := request.Request{W: res, Req: req}
+	qclass := state.QClass()
+	qtype := state.QType()
+	qname := state.Name()
+
+	if qclass == dns.ClassINET && (qtype == dns.TypeA || qtype == dns.TypeAAAA) {
+		ips, err := sd.sdcClient.Discover(qname)
+		if err != nil {
+			sd.log.Error(err)
+			return dns.RcodeServerFailure, err
+		}
+
+		if len(ips) > 0 {
+			res = &ServiceDiscoveryResponseWriter{
+				ResponseWriter: res,
+				ips:            ips,
+				ttl:            sd.ttl,
+			}
+		}
+	}
+
+	return plugin.NextOrFailure(pluginName, sd.Next, ctx, res, req)
+}
+
+// ServiceDiscoveryResponseWriter wraps a dns.ResponseWriter so it can respond
+// with discovered service addresses.
+type ServiceDiscoveryResponseWriter struct {
+	dns.ResponseWriter
+
+	ips []string
+	ttl uint32
+}
+
+// WriteMsg writes the response for the discovered service.
+func (sdrw *ServiceDiscoveryResponseWriter) WriteMsg(res *dns.Msg) error {
+	answer := make([]dns.RR, 0)
+	qtype := res.Question[0].Qtype
```

Comment:
What if there was more than one question? (Is that a thing? The structure looks like an array, so maybe?)

Reply:
Multiple questions are not a reality. You can look up online and you will repeatedly get the same answer: nobody practices this and it's not well defined how the behaviour should be.